singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,366 @@
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(Vinet)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"Vinet Params: "};
+    printf("%s rho0:%e T0:%e B0:%e BP0:%e\n  A0:%e Cv0:%e E0:%e S0:%e\n",
+           st, _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0);
+    for( int i=0; i<39; i++) {
+	    if(_d2to40[i] > 0.0) printf("d%i:%e\t",i+2,_d2to40[i]);
+    }
+    printf("\n\n");
+  }
+  //Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Vinet"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  //AEM: I am not sure what this does so leave it in.
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::temperature;
+  Real _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0;
+  const Real *_expconsts;
+  Real _VIP[43], _d2to40[39];
+  PORTABLE_INLINE_FUNCTION void CheckVinet();
+  PORTABLE_INLINE_FUNCTION void Vinet_F_DT_func(const Real rho, const Real T, 
+		  				Real *output) const;
+};
+
+PORTABLE_INLINE_FUNCTION void Vinet::CheckVinet() {
+
+  int ind;
+
+  if (_rho0 < 0.0){
+    printf("testing rho0\n");
+    EOS_ERROR("Required Vinet model parameter rho0 not set");
+    return;
+  }
+  if (_T0 < 0.0){
+    printf("testing T0\n");
+    EOS_ERROR("Required Vinet model parameter T0 not set");
+    return;
+  }
+  if (_B0 < 0.0){
+    printf("testing B0\n");
+    EOS_ERROR("Required Vinet model parameter B0 not set");
+    return;
+  }
+  if (_BP0 < 1.0){
+    printf("testing BP0\n");
+    EOS_ERROR("Required Vinet model parameter BP0 not set or too small (<1)");
+    return;
+  }
+  if (_A0 < 0.0){
+    printf("testing A0\n");
+    EOS_ERROR("Required Vinet model parameter A0 not set");
+    return;
+  }
+  if (_Cv0 < 0.0){
+    printf("testing Cv0\n");
+    EOS_ERROR("Required Vinet model parameter Cv0 not set");
+    return;
+  }
+
+
+  for(ind = 0; ind < 39; ind++) _d2to40[ind]=_expconsts[ind];
+
+  _VIP[0] = _B0/_rho0 + pow(_A0*_B0/_rho0,2.0)*_T0/_Cv0;      // sound speed squared
+  _VIP[1] = 3.0/2.0*(_BP0-1.0); // exponent
+  _VIP[2] = 1.0; // prefactor f0
+  _VIP[3] = 0.0; // prefactor f1
+
+  for (ind = 4; ind < 43; ind++) _VIP[ind] = _d2to40[ind-4];
+  ind = 42;
+  while (_VIP[ind] == 0.0) ind--;
+  while (ind >= 4) {
+    ind--;
+    _VIP[ind] = _VIP[ind] - (ind -1)/_VIP[1]*_VIP[ind+1];
+  }
+}
```

Comment:
I'm always really wary of using `while` loops when the length of an array is known. I think it makes the code less intuitive and I feel a `for` loop can do just as well. Moreover, my read of this code is that the equation reduces to
```c++
_VIP[ind] = _VIP[ind]
```
when `_VIP[ind+1] = 0`. To me this says that you can skip the first `while` loop and replace the second one with a `for` loop of the form
```c++
for (int i = 41; i >=4 ; i--) {
  _VIP[i] = _VIP[i] - (i - 1)/_VIP[1] * _VIP[i + 1];
}
```
Or better yet, you can avoid the aobe copy of `_expconsts` into `_d2to40` completely and just use

```c++
for (int i = 38; i >=0 ; i--) {
  _VIP[i + 4] = _expconsts[i] - (i - 1)/_VIP[1] * _expconsts[i + 1];
}
```

Reply:
I agree with @jhp-lanl here. Also have some comments about this pattern broadly, as well as a few nitpicks about code safety.